README.md
<a id="x-28DOCS-2FDOCS-3A-40README-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

# GitHub Action to Build Documentation for a Common Lisp Library

This is a Github Action can be used to build docs and update `gh-pages` branch used to
host static site using [GitHub Pages][9f4c].

It should be used after the [setup-lisp][8de1] action.

<a id="x-28DOCS-2FDOCS-3A-3A-40FEATURES-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## What this action does for you?

* It runs [docs-builder][f2be] to build `HTML` docs.

* If any there are any changes, it uploads them into the `gh-pages` branch.

* Also, it pushes any other changes to the current branch. This way, `README`.md
  is updated when you are using [MGL-PAX][ecfb]
  or [40ANTS-DOC][f503]
  as a docs-builder.

<a id="x-28DOCS-2FDOCS-3A-3A-40TYPICAL-USAGE-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## A typical usage

Here is how a minimal GitHub Workflow might look like.

Create a `.github/workflows/docs.yml` file with following content:

```yaml
name: 'Docs'

on:
  # This will run tests on pushes
  # to master branch and every monday:
  push:
    branches:
      - 'main'
      - 'master'
  schedule:
    - cron:  '0 10 * * 1'

jobs:
  build-docs:
    runs-on: ubuntu-latest
    
    env:
      LISP: sbcl-bin

    steps:
      - uses: actions/checkout@v1
      - uses: 40ants/setup-lisp@v1
      - uses: 40ants/build-docs@v1
        with:
          asdf-system: my-system
```
The part, corresponding to an action call is:

```yaml
- uses: 40ants/build-docs@v1
  with:
    asdf-system: my-system
```
Here we provided a system name `my-system`, and
action will figure out how to build it docs, because
it uses smart guesser from
[docs-builder][f2be].

You should check if the documentation builder is supported
by `docs-builder` and if not, you can easily to add this support.

<a id="x-28DOCS-2FDOCS-3A-3A-40ROADMAP-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## Roadmap

* Add more documentation builders to `docs-builder`.

* Vendor all dependencies, to make action more reliable and secure.


[f2be]: https://40ants.com/docs-builder/
[8de1]: https://40ants.com/setup-lisp/
[f503]: https://github.com/40ants/doc
[ecfb]: https://github.com/cl-doc-systems/mgl-pax
[9f4c]: https://pages.github.com/

* * *
###### [generated by [40ANTS-DOC](https://40ants.com/doc/)]


ChangeLog.md
<a id="x-28DOCS-2FCHANGELOG-3A-40CHANGELOG-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

# ChangeLog

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E3-2E3-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.3.3 (2022-11-12)

Now action outputs more debug information when you rerun it with "Enable debug logging" option.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E3-2E2-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.3.2 (2022-11-11)

Fixed how does .qlot get updated. Now we are running `qlot update --no-deps`.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E3-2E1-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.3.1 (2022-08-01)

Don't push docs to the github if last commit is last documentation update
(has commit message is "Update docs").

This way you can set up a scheduled execution for the workflow, and
 it will not fill git repository with endless updates.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E3-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.3.0 (2021-10-19)

Don't upload docs when action is used on pull-request.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E2-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.2.0 (2021-05-10)

Now action will check if `log4cl-extras`, `docs-builder` or `ngrok`
are present in `qlfile`.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E1-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.1.0 (2021-04-15)

Now builder will be runned in the mode when any warning
is considered as an error. To prevent this behaviour,
pass `error-on-warnings: false` argument. 

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E0-2E7-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.0.7 (2021-04-05)

Install `log4cl-extras` because docs-builder now depends on it
and this project is not in the Quicklisp yet.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E0-2E6-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.0.6 (2021-04-05)

Switched to the [`40ants-doc`][a2c7] for its own documentation.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E0-2E5-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.0.5 (2021-03-08)

Fixed the case when during the docs building
some file in the `.github/workflows` has changed.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E0-2E4-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.0.4 (2021-02-28)

Fixed update of the qlfile in case there is no new line at the end.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E0-2E3-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.0.3 (2021-02-24)

Fixed utilities running when cache is enabled.
We need to call them under the qlot.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E0-2E2-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.0.2 (2021-02-16)

Yet another fix of pushing changes into the current branch on push event.

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E0-2E1-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.0.1 (2021-02-16)

Fixed error when we are uploading local changes on "push event".

<a id="x-28DOCS-2FCHANGELOG-3A-3A-7C1-2E0-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 1.0.0 (2021-01-07)

Initial version.


[a2c7]: https://40ants.com/doc/#x-28-23A-28-2810-29-20BASE-CHAR-20-2E-20-2240ants-doc-22-29-20ASDF-2FSYSTEM-3ASYSTEM-29

* * *
###### [generated by [40ANTS-DOC](https://40ants.com/doc/)]


docs.lisp
(defpackage #:docs/docs
  (:use #:cl)
  (:import-from #:40ants-doc
                #:defsection
                #:defsection-copy)
  (:import-from #:docs-config
                #:docs-config)
  (:import-from #:docs/changelog
                #:@changelog)
  (:export #:@index
           #:@readme
           #:@changelog))
(in-package docs/docs)


(defmethod docs-config ((system (eql (asdf:find-system "docs"))))
  ;; 40ANTS-DOC-THEME-40ANTS system will bring
  ;; as dependency a full 40ANTS-DOC but we don't want
  ;; unnecessary dependencies here:
  (ql:quickload :40ants-doc-theme-40ants)
  (list :theme
        (find-symbol "40ANTS-THEME"
                     (find-package "40ANTS-DOC-THEME-40ANTS"))))


(defsection @index (:title "GitHub Action to Build Documentation for a Common Lisp Library"
                    :ignore-words ("HTML"
                                   "MGL-PAX"
                                   "README")
                    :external-docs ("https://40ants.com/doc/"))
  "
This is a Github Action can be used to build docs and update `gh-pages` branch used to
host static site using [GitHub Pages](https://pages.github.com/).

It should be used after the [setup-lisp](https://40ants.com/setup-lisp/) action.
"
  (@features section)
  (@typical-usage section)
  (@roadmap section))


(defsection-copy @readme @index)


(defsection @features (:title "What this action does for you?")
  "
* It runs [docs-builder](https://40ants.com/docs-builder/) to build HTML docs.
* If any there are any changes, it uploads them into the `gh-pages` branch.
* Also, it pushes any other changes to the current branch. This way, README.md
  is updated when you are using [MGL-PAX](https://github.com/cl-doc-systems/mgl-pax)
  or [40ANTS-DOC](https://github.com/40ants/doc)
  as a docs-builder.
")


(defsection @typical-usage (:title "A typical usage")
  "
Here is how a minimal GitHub Workflow might look like.

Create a `.github/workflows/docs.yml` file with following content:

```yaml
name: 'Docs'

on:
  # This will run tests on pushes
  # to master branch and every monday:
  push:
    branches:
      - 'main'
      - 'master'
  schedule:
    - cron:  '0 10 * * 1'

jobs:
  build-docs:
    runs-on: ubuntu-latest
    
    env:
      LISP: sbcl-bin

    steps:
      - uses: actions/checkout@v1
      - uses: 40ants/setup-lisp@v1
      - uses: 40ants/build-docs@v1
        with:
          asdf-system: my-system
```

The part, corresponding to an action call is:

```yaml
- uses: 40ants/build-docs@v1
  with:
    asdf-system: my-system
```

Here we provided a system name `my-system`, and
action will figure out how to build it docs, because
it uses smart guesser from
[docs-builder](https://40ants.com/docs-builder/).

You should check if the documentation builder is supported
by `docs-builder` and if not, you can easily to add this support.
")


(defsection @roadmap (:title "Roadmap")
  "
- Add more documentation builders to `docs-builder`.
- Vendor all dependencies, to make action more reliable and secure.
")


changelog.lisp
(defpackage #:docs/changelog
  (:use #:cl)
  (:import-from #:40ants-doc/changelog
                #:defchangelog))
(in-package docs/changelog)


(defchangelog ()
  (1.3.3 2022-11-12
         "Now action outputs more debug information when you rerun it with \"Enable debug logging\" option.")
  (1.3.2 2022-11-11
         "Fixed how does .qlot get updated. Now we are running `qlot update --no-deps`.")
  (1.3.1 2022-08-01
         "Don't push docs to the github if last commit is last documentation update
        (has commit message is \"Update docs\").

         This way you can set up a scheduled execution for the workflow, and
         it will not fill git repository with endless updates.")
  (1.3.0 2021-10-19
         "Don't upload docs when action is used on pull-request.")
  (1.2.0 2021-05-10
         "Now action will check if `log4cl-extras`, `docs-builder` or `ngrok`
          are present in `qlfile`.")
  (1.1.0 2021-04-15
         "Now builder will be runned in the mode when any warning
          is considered as an error. To prevent this behaviour,
          pass `error-on-warnings: false` argument. ")
  (1.0.7 2021-04-05
         "Install `log4cl-extras` because docs-builder now depends on it
          and this project is not in the Quicklisp yet.")
  (1.0.6 2021-04-05
         "Switched to the 40ANTS-DOC for its own documentation.")
  (1.0.5 2021-03-08
         "Fixed the case when during the docs building
          some file in the `.github/workflows` has changed.")
  (1.0.4 2021-02-28
         "Fixed update of the qlfile in case there is no new line at the end.")
  (1.0.3 2021-02-24
         "Fixed utilities running when cache is enabled.
          We need to call them under the qlot.")
  (1.0.2 2021-02-16
         "Yet another fix of pushing changes into the current branch on push event.")
  (1.0.1 2021-02-16
         "Fixed error when we are uploading local changes on \"push event\".")
  (1.0.0 2021-01-07
         "Initial version."))
